{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# What is the candidate script?  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The candidate script (pedigree_impala_analysis.py) is an annotation program that allows a user to analyze variants in a pedigree based upon a particular inheritance mode. The user can use one or more inheritance modes to run the script, control results by providing various thresholds in the user specified region (jupyter) or in the configuration file. The candidate script also annotates a variant with a stress score, which is obtained by processing the target inheritance pattern and the observed pattern w.r.t to an allele (Segregation class). The variant information is loaded into platform specific tables on the p7_platform database on impala. Current Annotation include: Kavair, CMS, UCSC_known_genes, CADD, Clinvar and SNPEff(depending upon availability on Impala). \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#!/tools/bin/python\n",
    "from __future__ import print_function \n",
    "from __future__ import division\n",
    "from impala.dbapi import connect\n",
    "from segregation import segregation\n",
    "from collections import Counter\n",
    "import datetime\n",
    "import sys\n",
    "import os.path\n",
    "\n",
    "#records timing\n",
    "today = datetime.date.today()\n",
    "target_hash={}\n",
    "segregation_object=segregation() #Object for the segregation class\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User specified region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Connection successful!\n"
     ]
    }
   ],
   "source": [
    "stress_cutoff=0.49\n",
    "kaviar_cutoff=0.03\n",
    "score_cutoff=3.5\n",
    "subject_list=['103-00003-02','103-00003-01','103-00003-03']\n",
    "target_inheritance=['aaaaab','abaaab','aaabab','ababbb']\n",
    "db_user=''\n",
    "restrict=''\n",
    "outdir=\" \"\n",
    "\n",
    "if not outdir:\n",
    "    print(\"Please enter location to an outdir\")\n",
    "\n",
    "##Database connection to Impala\n",
    "try:\n",
    "    conn=connect(host='glados19', port=21050,database=db_user)\n",
    "    DB = conn.cursor()\n",
    "    DB.execute(\"SELECT VERSION()\")\n",
    "    results=DB.fetchone()\n",
    "    print(\"Connection successful!\")\n",
    "except:\n",
    "    print(\"Connection failed\")\n",
    "    sys.exit()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created separate ouput files for each inheritance pattern\n"
     ]
    }
   ],
   "source": [
    "##header_out add output headers to the output file\n",
    "def header_out(pattern,writer):\n",
    "    writer.write('#stress_cutoff: '+ str(stress_cutoff) + \"\\n\")\n",
    "    writer.write('#kaviar_cutoff: '+ str(kaviar_cutoff) + \"\\n\")\n",
    "    writer.write('#score_cutoff: ' + str(score_cutoff) + \"\\n\")\n",
    "    writer.write('#Inclusion_order: ' + str(subject_list) + \"\\n\")\n",
    "    writer.write('#target_pattern: ' + pattern + \"\\n\")\n",
    "    writer.write('#date:time: ' + str(datetime.datetime.now()) + \"\\n\")\n",
    "    writer.write(\"#Chromosome\\tPosition\\tReference\\tCandidate\\tGenotypes\\tGenotype_vectors\\tCandidate_stress\\tAllele_count_in_pedigree\\tMax_score\\tMin_score\\tAverage_score\\tKaviar\\tDANN\\tCMS\\tClinvar_sig\\tGene_name\\tOverall_candidate_score\" + \"\\n\")\n",
    "\n",
    "#Creates dict/hash for all target patterns\n",
    "\n",
    "for pattern in target_inheritance:\n",
    "    completeName = os.path.join(outdir, 'candidate_pattern_'+pattern + '.txt')\n",
    "    writer_pattern = open(completeName, 'w')\n",
    "    target_hash[pattern]=writer_pattern\n",
    "    header_out(pattern,writer_pattern)\n",
    "print (\"Created separate ouput files for each inheritance pattern\")\n",
    "    \n",
    "\n",
    "\"\"\"\n",
    "Vividict allows creation of perl like\n",
    "hashes of hashes or multi key level hash\n",
    "User would always have to point their hash/dict to Vividict\n",
    "Eg: hash_dict=Vividict()\n",
    "\n",
    "\"\"\"\n",
    "class Vividict(dict):\n",
    "    def __missing__(self, key):\n",
    "        value= self[key] = type(self)()\n",
    "        return value\n",
    "\n",
    "\"\"\"\n",
    "The compute_sequence_quality_score_statistics returns\n",
    "max,min and average scores seen at a position.\n",
    "\"\"\"\n",
    "def compute_sequence_quality_score_statistics(score_array):\n",
    "    max_score=max(score_array)\n",
    "    min_score=min(score_array)\n",
    "    num_list=[float(x) for x in score_array]\n",
    "    avg_score=sum(num_list)/len(num_list)\n",
    "    return (max_score,min_score,avg_score)\n",
    "\n",
    "\"\"\"\n",
    "QAF_score_modifier provides returns numerical\n",
    "results based upon the allele frequenies that are\n",
    "supplied. This score is essential in ranking the \n",
    "candidates or for calculating the candidate score\n",
    "\"\"\"\n",
    "def QAF_score_modifier(QAF): \n",
    "    if QAF == 0:\n",
    "        return 1\n",
    "    if QAF < 0.00001:\n",
    "        return 1/0.99\n",
    "    if QAF < 0.0001:\n",
    "        return 1/0.98\n",
    "    if QAF < 0.01:\n",
    "        return 1/0.9\n",
    "    if QAF < 0.05:\n",
    "        return 1/0.75\n",
    "    if QAF < 0.1:\n",
    "        return 1/0.5\n",
    "    if QAF < 0.15:\n",
    "        return 1/0.3\n",
    "    return 1\n",
    "\"\"\"\n",
    "This returns a numerical value based upon\n",
    "what the score value is supplied.\n",
    "The returned value is essential in calculating candidate score\n",
    "\"\"\"\n",
    "def quality_score_adjustment(max_score):\n",
    "    score = max_score\n",
    "    return_value=0\n",
    "    \n",
    "    if score >=50:\n",
    "        return 0  #anything above 50 is good\n",
    "    elif score > 35:\n",
    "        return_value=(50-score)/15\n",
    "        return return_value\n",
    "        #minimum score is 20\n",
    "    else:\n",
    "        return 4 \n",
    "\n",
    "\"\"\"\n",
    "This is similiar to above score modifier.\n",
    "The motivation behind this function is to return low value\n",
    "for highly pathogenic variants.\n",
    "\"\"\"\n",
    "\n",
    "def dann_score_modifier(dann_score):\n",
    "    if dann_score >= 0.995:\n",
    "        return 0.1\n",
    "    elif dann_score >= 0.98 and dann_score < 0.995:\n",
    "        return 0.5\n",
    "    elif dann_score >= 0.93 and dann_score < 0.98:\n",
    "        return 0.7\n",
    "    else:\n",
    "        return 1\n",
    "\"\"\"\n",
    "The overall candidate score is a rank given to a variant based \n",
    "upon inputs of genetic_stress, QAF,DANN, CMS etc.\n",
    "Lower candidate score means that a variant is likely to be more pathogenic\n",
    "\"\"\"\n",
    "def overall_candidate_score(genetic_analysis_stress,queried_allele_frequency,dann_score,max_score,cms):\n",
    "    QAF_score_modified=QAF_score_modifier(queried_allele_frequency)\n",
    "    modified_dann=dann_score_modifier(dann_score)\n",
    "    score = ((genetic_analysis_stress + 1) * (1 + QAF_score_modified) * (1 + cms) * modified_dann)\n",
    "    adjusted_quality_score = quality_score_adjustment(max_score)\n",
    "    overall_score = score * (1 + adjusted_quality_score)\n",
    "    return(overall_score) \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process Allele and quality score\n",
    "process_alleles is the heart of the pedigree_analysis program.\n",
    "1)This re-creates family specific genotypes. \n",
    "2)Processes genotype quality\n",
    "3)Creates dict/hash to store kaviar,dann,gene and cms annotation.\n",
    "4)Annotates individual variants \n",
    "5)Calls the overall candidate score\n",
    "6)Filters the output based upon user specified thresholds.\n",
    "Currently,genotypes with GT ./. or 0/0 or missing from the impala database\n",
    "are converted to homozygous reference calls."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def process_alleles(list):\n",
    "    kaviar={}\n",
    "    quality={}\n",
    "    dann={}\n",
    "    clinvar={}\n",
    "    subject=Vividict()\n",
    "    GT={}\n",
    "    chromosome=''\n",
    "    position=''\n",
    "    reference=''\n",
    "    cms=''\n",
    "    gene={}\n",
    "    for line in list:\n",
    "        chromosome=line[1]\n",
    "        position=line[2]\n",
    "        reference=line[3]\n",
    "        subject[line[0]][line[4]]=1\n",
    "        GT[line[0]]=line[5]\n",
    "        try:\n",
    "            if line[7] is not None:\n",
    "                kaviar[line[4]]=line[7]\n",
    "                if kaviar[line[4]] != line[7]:\n",
    "                    kaviar[line[4]]+=line[7]\n",
    "        except KeyError:\n",
    "            kaviar[line[4]]=0\n",
    "        \n",
    "        if chromosome=='M':\n",
    "            dann['A']=0.1\n",
    "            dann['T']=0.1\n",
    "            dann['G']=0.1\n",
    "            dann['C']=0.1\n",
    "        elif chromosome=='M' and stress_cutoff >= 2.0:\n",
    "            dann['A']=1\n",
    "            dann['T']=1\n",
    "            dann['G']=1\n",
    "            dann['C']=1\n",
    "        else:\n",
    "            dann['A']=line[8]\n",
    "            dann['T']=line[9]\n",
    "            dann['G']=line[10]\n",
    "            dann['C']=line[11]\n",
    "\n",
    "        if line[6] is None:\n",
    "            quality[0]=1\n",
    "        else:\n",
    "            quality[line[6]]=1\n",
    "        clinvar[line[4]]=line[13] \n",
    "        cms=line[12]\n",
    "        if line[14] is not None:\n",
    "            gene[line[14]]=1\n",
    "    \n",
    "    quality_list=[]\n",
    "    \n",
    "    for gqx in quality:\n",
    "        quality_list.append(gqx)\n",
    "    \n",
    "    quality_array=compute_sequence_quality_score_statistics(quality_list)    \n",
    "    max_score=quality_array[0]\n",
    "    average_score=\"%.0f\" % (quality_array[2])\n",
    "    min_score=quality_array[1]    \n",
    "      \n",
    "    maximum=max(dann, key=dann.get)  \n",
    "    dann['max']=dann[maximum]\n",
    "\n",
    "    freq=0\n",
    "    for value in kaviar:\n",
    "        freq += kaviar[value]\n",
    "\n",
    "    reference_freq=(1.0 - freq)    \n",
    "    kaviar[reference]=reference_freq\n",
    "\n",
    "    gene_string=' '.join(gene.keys())\n",
    "    allele_string=''\n",
    "    allele_array=[]\n",
    "    for member in subject_list:\n",
    "        if member in subject:\n",
    "            if member in GT:\n",
    "                if GT[member]==\"1/1\":\n",
    "                    alt=subject[member].keys()\n",
    "                    alt=''.join(alt)\n",
    "                    allele_array.append(alt)\n",
    "                    allele_array.append(alt)\n",
    "                    allele_string += alt\n",
    "                    allele_string += ' '\n",
    "                    allele_string += alt\n",
    "                    allele_string += ' '\n",
    "                elif GT[member]==\"0/1\":\n",
    "                    alt=subject[member].keys()\n",
    "                    alt=''.join(alt)\n",
    "                    allele_array.append(reference)\n",
    "                    allele_array.append(alt)\n",
    "                    allele_string += reference\n",
    "                    allele_string += ' '\n",
    "                    allele_string += alt\n",
    "                    allele_string += ' '\n",
    "                elif GT[member]==\"1/0\":\n",
    "                    alt=subject[member].keys()\n",
    "                    alt=''.join(alt)\n",
    "                    allele_array.append(alt)\n",
    "                    allele_array.append(reference)\n",
    "                    allele_string += alt\n",
    "                    allele_string += ' '\n",
    "                    allele_string += reference\n",
    "                    allele_string += ' '\n",
    "                elif GT[member]==\"1/2\" or GT[member]==\"2/1\":\n",
    "                    alt = subject[member].keys()\n",
    "                    for allele in alt:\n",
    "                        allele_array.append(allele)\n",
    "                        allele_string += allele\n",
    "                        allele_string += ' '\n",
    "               \n",
    "        else:\n",
    "            allele_array.append(reference)\n",
    "            allele_array.append(reference)\n",
    "            allele_string += reference\n",
    "            allele_string += ' '\n",
    "            allele_string += reference\n",
    "            allele_string += ' '\n",
    "    allele_counter=Counter(allele_array)   \n",
    "    Unique_array=set(allele_array)\n",
    "    \n",
    "    for candidate in Unique_array:\n",
    "        for target_pattern in target_inheritance:\n",
    "            genotype_vectors=segregation_object.standardized_genotype_vector_with_reference_to_a_particular_allele(allele_array,candidate)\n",
    "            candidate_stress=segregation_object.target_test(genotype_vectors,target_pattern)\n",
    "            if candidate_stress=='NA':\n",
    "                continue\n",
    "            kaviar_score=0\n",
    "            dann_score=0\n",
    "            cms_range=''\n",
    "            clin_sig=''\n",
    "            if candidate in kaviar:\n",
    "                kaviar_score=kaviar[candidate]\n",
    "            else:\n",
    "                kaviar_score=0\n",
    "     \n",
    "            try:\n",
    "                if dann[candidate] is not None:\n",
    "                    dann_score=\"%4f\" % dann[candidate]\n",
    "            except KeyError: \n",
    "                dann_score=0\n",
    "         \n",
    "            if len(reference) > 1 or len(candidate) > 1:\n",
    "                dann_score=dann['max']\n",
    "                if dann_score:\n",
    "                    dann_score=\"%4f\" % (dann_score)\n",
    "            elif candidate==reference:\n",
    "                dann_score=dann['max']\n",
    "                if dann_score:\n",
    "                    dann_score=\"%4f\" % (dann_score)\n",
    "       \n",
    "            if candidate in clinvar:\n",
    "                if clinvar[candidate] is None:\n",
    "                    clin_sig=0\n",
    "                else:\n",
    "                    clin_sig=clinvar[candidate]\n",
    "            else:\n",
    "                clin_sig=0\n",
    "        \n",
    "            if cms is None:     \n",
    "                cms_range=0\n",
    "            else:\n",
    "                cms_range=1\n",
    "            \n",
    "            output=[]   \n",
    "            overall_candidate_score_value=0                \n",
    "            overall_candidate_score_value=overall_candidate_score(candidate_stress,kaviar_score,float(dann_score),max_score,cms_range)\n",
    "            overall_candidate_score_value= float(overall_candidate_score_value)\n",
    "            if kaviar_cutoff >= float(kaviar_score):\n",
    "                if stress_cutoff >= candidate_stress:\n",
    "                    if score_cutoff >=overall_candidate_score_value:\n",
    "                        if kaviar_score==0:\n",
    "                            kaviar_score=0\n",
    "                        elif kaviar_score < 0.0001:\n",
    "                            kaviar_score=\"%2E\" % (kaviar_score)\n",
    "                            kaviar_score=\"%4f\" %(float(kaviar_score))\n",
    "                        else:\n",
    "                            kaviar_score=\"%4f\" %(kaviar_score)\n",
    "                        overall_candidate_score_value= \"%4f\" %(overall_candidate_score_value)\n",
    "                        output=[chromosome,position,reference,candidate,allele_string,genotype_vectors,candidate_stress,allele_counter[candidate],max_score,min_score,average_score,kaviar_score,dann_score,cms_range,clin_sig,gene_string,overall_candidate_score_value]\n",
    "                        for lines in output:\n",
    "                            target_hash[target_pattern].write(str(lines) + \"\\t\")\n",
    "                        target_hash[target_pattern].write(\"\\n\")    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotation joins\n",
    "\n",
    "This is reponsible for running chromosome sepcific left outer joins\n",
    "between the temp tables and annotation tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def query_impala(chr):\n",
    "    mapper=Vividict()\n",
    "    counter=0\n",
    "    list=[]\n",
    "    \n",
    "    try:\n",
    "        query=(\"select f.*, kav.allele_freq,dann.score_a as A,dann.score_t as T,\"\n",
    "           \"dann.score_g as G,dann.score_c as C, cms.start,clin.clin_sig,ucsc.gene_name from \"\n",
    "          \"(select fam.subject_id, fam.chrom, fam.pos,fam.ref,fam.alt,fam.gt,fam.gq\"  \n",
    "          \" from %s.temp as fam \"\n",
    "          \"where fam.chrom='%s') as f \"\n",
    "          \"left outer join p7_ref_grch37.kaviar_isb as kav \"\n",
    "          \"on f.chrom=kav.chrom \"\n",
    "          \"and f.pos=kav.pos \"\n",
    "          \"and f.ref=kav.ref \"\n",
    "          \"and f.alt=kav.alt \"\n",
    "          \"left outer join p7_ref_grch37.dann as dann \"\n",
    "          \"on f.chrom=dann.chrom \"\n",
    "          \"and f.pos=dann.pos \"\n",
    "          \"left outer join p7_itmi.cms_gt1 as cms \"\n",
    "          \"on concat('chr',f.chrom)=cms.chrom \"\n",
    "          \"and f.pos >= cms.start and f.pos <= cms.stop \"\n",
    "          \"left outer join p7_ref_grch37.clinvar as clin \"\n",
    "          \"on f.chrom=clin.chrom \"\n",
    "          \"and f.pos=clin.pos \"\n",
    "          \"and f.ref=clin.ref \"\n",
    "          \"and f.alt=clin.alt \"\n",
    "          \"left outer join p7_ref_grch37.ucsc_genes as ucsc \"\n",
    "          \"on f.chrom=ucsc.chrom \"\n",
    "          \"and f.pos >= ucsc.txstart and f.pos <= ucsc.txend \"          \n",
    "          \"order by f.pos\" % (db_user,chr)) \n",
    "        DB.execute(query)\n",
    "        print (\"Finished running annotation joins on chrom \" + str(chr),end=\"\\n\")\n",
    "\n",
    "    except:\n",
    "        print(\"Join failed, please check if impala is online and all tables are present\",end=\"\\n\")\n",
    "        print(\"Since joined has failed I am dropping the temp table\",end=\"\\n\")\n",
    "        sys.exit()\n",
    "        \n",
    "    for row in DB:\n",
    "        subject_id = row[0]\n",
    "        chrom = row[1]\n",
    "        pos = row[2] \n",
    "        if mapper[chrom][pos]:\n",
    "            list.append(row)\n",
    "            counter+=1\n",
    "        else: \n",
    "            if len(list) > 0:\n",
    "                process_alleles(list)\n",
    "            list=[]\n",
    "            mapper[chrom][pos]=1\n",
    "            list.append(row)\n",
    "    \n",
    "    process_alleles(list) \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating temprorary table\n",
    "Creates a temporary table so that annotation joins could be done\n",
    "This is created in db_user database as temp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def create_table():\n",
    "    print (\"Creating a temp table on database \" + db_user,end=\"\\n\")\n",
    "    \n",
    "    query_string=''\n",
    "    for string in subject_list:\n",
    "        query_string += \"'\" + string + \"'\" + \",\"\n",
    "    query_string=query_string[:-1]\n",
    "    \n",
    "    try:\n",
    "        DB.execute((\"DROP TABLE IF EXISTS %s.temp\") %(db_user))\n",
    "        query=(\"create table %s.temp as\"\n",
    "              \"(select * from p7_platform.wgs_illumina_variant where subject_id IN (%s))\" % (db_user,query_string))\n",
    "        DB.execute(query) \n",
    "        print (\"Temporary table created on Database \" + db_user,end=\"\\n\" )\n",
    "    except:\n",
    "        print(\"Couldn't create a temporary table are you sure the database information is correct\",end=\"\\n\")\n",
    "        sys.exit()\n",
    "    \n",
    "    if restrict:\n",
    "        for chr in restrict:\n",
    "            query_impala(chr)\n",
    "    else:\n",
    "        chrArray=['1','2','3','4','5','6','7','8','9','10','11','12','13','14','15','16','17','18','19','20','21','22','X','Y']\n",
    "        for chr in chrArray:\n",
    "            query_impala(chr)\n",
    "\n",
    "    drop_table()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Deleting temporary table\n",
    "Once the analysis done the temp table is dropped."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def drop_table():\n",
    "    try:\n",
    "        query=(\"drop table %s.temp\" %(db_user))\n",
    "        DB.execute(query)\n",
    "        print (\"Analysis is complete deleting temporary table\",end=\"\\n\")\n",
    "    except:\n",
    "        print (\"Failed to delete temporary table\",end=\"\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Program starts and creates temprorary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating a temp table on database users_hrishi\n",
      "Temporary table created on Database users_hrishi\n",
      "Finished running annotation joins on chrom 1\n",
      "Analysis is complete deleting temporary table\n"
     ]
    }
   ],
   "source": [
    "create_table()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
